dialer.go

```diff
@@ -68,175 +88,178 @@ func NewDialer(opts *DialerOpts) Dialer {
-	windowSize            int
-	maxPadding            int
-	maxLiveConns          int
-	maxStreamsPerConn     uint16
-	idleInterval          time.Duration
-	pingInterval          time.Duration
-	redialSessionInterval time.Duration
-	pool                  BufferPool
-	cipherCode            Cipher
-	serverPublicKey       *rsa.PublicKey
-	muNumLivePending      sync.Mutex
-	numLive               int
-	numPending            int
-	liveSessions          chan sessionIntf
-	emaRTT                *ema.EMA
+	// We create another background session with a shorter dial timeout to ensure liveness in the
+	// case of network disruptions.
+	d.pendingSessions <- &sessionConfig{
+		name:         "liveness to " + d.name,
+		dialTimeout:  time.Duration(opts.ShortDialTimeout) * time.Second,
+		sleepOnError: 1 * time.Second,
+	}
+	return d
 }
 
-func (d *dialer) Dial(dial DialFN) (net.Conn, error) {
-	return d.DialContext(context.Background(), dial)
+// maintainTCPConnections maintains background TCP connection(s) and associated lampshade session(s)
+func (d *dialer) maintainTCPConnections() {
+	for sc := range d.pendingSessions {
+		go d.trySession(sc)
+	}
 }
 
-func (d *dialer) DialContext(ctx context.Context, dial DialFN) (net.Conn, error) {
-	s, err := d.getOrCreateSession(ctx, dial)
+func (d *dialer) trySession(sc *sessionConfig) {
+	start := time.Now()
+	s, err := d.startSession(sc)
 	if err != nil {
-		return nil, err
+		d.log.Debugf("Error starting session '%v': %v", sc.name, err.Error())
+		// Sleeping when there's an error is necessary particularly for the case where the network is
+		// down, as this will then spin endlessly.
+		time.Sleep(sc.sleepOnError)
+		d.pendingSessions <- sc
+	} else {
+		d.log.Debugf("Created session in %v to %#v", time.Since(start), sc)
+		d.recycleSession(s)
 	}
-	c := s.CreateStream()
-	d.returnSession(s)
-	return c, nil
 }
 
-func (d *dialer) getNumLivePending() int {
-	d.muNumLivePending.Lock()
-	numLivePending := d.numLive + d.numPending
-	d.muNumLivePending.Unlock()
-	return numLivePending
+func (d *dialer) recycleSession(s sessionIntf) {
+	if !s.allowNewStream(d.maxStreamsPerConn) {
+		d.log.Debugf("Maximum streams reached for session to %v", d.name)
+		// The default number of streams per session is 65535, so this is unlikely to be reached. If it
+		// is, we create an additional session with the same configuration as the full session.
+		d.pendingSessions <- s.getSessionConfig()
+		return
+	}
+	// We now have a new or established TCP connection/session. At this point two things can happen:
+	// 1) The connection can be closed for any reason, in which case we want to request a new one
+	// 2) A dialer can request the session. In that case, it will retrieve the session.
+	//
+	// Selecting on those two cases avoids closed sessions building up in the live session channel.
+	select {
+	case <-s.getCloseCh():
+		d.log.Debugf("Session closed before requested. Requesting new session: %#v", s.getSessionConfig())
```

Comment:
I got around 10 such messages after running for 4 minutes, suggested that some sessions were wasted. Checking in detail...

Reply:
Those were idle closed. Not new, but better to avoid creating sessions if the dialer is never selected.

The current code avoided this by starting session on demand, with the cost of taking longer to switch to an idled proxy, because dialing a new session is required. OTOH, when more than one proxy look fail / get blocked at the same time, a pre-established session is nothing but an illusion.

Reply:
45 "Session closed before requested" when running for 30 minutes, of which 40 (30m / 1m30s * 2) were from a proxy never selected. Not that bad.